File: exercise/145/iteration.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> result;
        stack<pair<TreeNode*, int>> nodes;
        nodes.push(make_pair(root, 0));
        while (!nodes.empty()) {
            auto now = nodes.top();
            nodes.pop();
            TreeNode *p = now.first;
            if (!p) continue;
            int state = now.second;
            if (state == 0) {
                nodes.push(make_pair(p, 1));
                nodes.push(make_pair(p->right, 0));
                nodes.push(make_pair(p->left, 0));
            } else {
                result.push_back(p->val);
            }
        }
        return result;
    }
};

File: exercise/145/iteration.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        stack = [(root, 0)]
        while len(stack) > 0:
            now = stack.pop()
            if now[0] is None:
                continue
            if now[1] == 0:
                stack.append((now[0], 1))
                stack.append((now[0].right, 0))
                stack.append((now[0].left, 0))
            else:
                result.append(now[0].val)
        return result
